File: 53.maximum-subarray.js
/*
 * @lc app=leetcode id=53 lang=javascript
 *
 * [53] Maximum Subarray
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
  let maxSum = nums[0]
  let currSum = 0
  for (let num of nums) {
    currSum += num
    if (currSum > maxSum) maxSum = currSum
    if (currSum < 0) currSum = 0
  }
  return maxSum
};
// @lc code=end

File: cpp/53.maximum-subarray.cpp
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        // iteratively add each num
        // if curr sum < 0, discard prev nums
        // O(n) time O(k) space
        int maxSum = INT_MIN;
        int currSum = 0;
        for (int num : nums)
        {
            currSum += num;
            maxSum = max(maxSum, currSum);
            currSum = max(currSum, 0);
        }

        return maxSum;
    }
};
// @lc code=end
